aws/src/main/java/org/apache/iceberg/aws/LakeFormationAwsClientFactory.java

```diff
@@ -0,0 +1,221 @@
+    this.secretAccessKey = properties.get(AwsProperties.S3FILEIO_SECRET_ACCESS_KEY);
+    this.sessionToken = properties.get(AwsProperties.S3FILEIO_SESSION_TOKEN);
+  }
+
+  @Override
+  public S3Client s3() {
+    return S3Client.builder()
+        .endpointOverride(URI.create(s3Endpoint))
+        .credentialsProvider(buildS3CredentialProvider())
+        .region(region)
+        .httpClientBuilder(UrlConnectionHttpClient.builder())
+        .build();
+  }
+
+  @Override
+  public GlueClient glue() {
+    return GlueClient.builder()
+        .credentialsProvider(getAwsCredentialsProvider())
+        .region(region)
+        .httpClientBuilder(UrlConnectionHttpClient.builder())
+        .build();
+  }
+
+  @Override
+  public KmsClient kms() {
+    return KmsClient.builder()
+        .credentialsProvider(getAwsCredentialsProvider())
+        .region(region)
+        .httpClientBuilder(UrlConnectionHttpClient.builder())
+        .build();
+  }
+
+  @Override
+  public DynamoDbClient dynamo() {
+    return DynamoDbClient.builder()
+        .credentialsProvider(getAwsCredentialsProvider())
+        .region(region)
+        .httpClientBuilder(UrlConnectionHttpClient.builder())
+        .build();
+  }
+
+  private AwsCredentialsProvider buildS3CredentialProvider() {
+
+    GlueClient glue = glue();
+    String glueCatalogId = properties.get(AwsProperties.GLUE_CATALOG_ID);
+    String databaseName = properties.get(AwsProperties.LAKE_FORMATION_DB_NAME);
+    String tableName = properties.get(AwsProperties.LAKE_FORMATION_TABLE_NAME);
+
+    GetTableResponse response = glue.getTable(GetTableRequest.builder()
+        .catalogId(glueCatalogId)
+        .databaseName(databaseName)
+        .name(tableName)
+        .build());
+    boolean isRegisteredWithLakeFormation = response.table().isRegisteredWithLakeFormation() != null ?
+        response.table().isRegisteredWithLakeFormation() : Boolean.FALSE;
+
+    if (isRegisteredWithLakeFormation) {
+      return new LakeFormationCredentialsProvider(properties);
+    } else {
+      return getAwsCredentialsProvider();
+    }
+  }
+
+  private static Region region(Map<String, String> properties) {
+    String region = properties.containsKey(AwsProperties.LAKE_FORMATION_AWS_REGION) ?
+        properties.get(AwsProperties.LAKE_FORMATION_AWS_REGION) : null;
+    if (region == null || region.isEmpty()) {
+      throw new RuntimeException("Region is empty");
+    }
+    return Region.of(region);
+  }
+
+  private AwsCredentialsProvider getAwsCredentialsProvider() {
+    return StaticCredentialsProvider.create(AwsSessionCredentials.create(accessKeyId, secretAccessKey, sessionToken));
+  }
+
+  static class LakeFormationCredentialsProvider implements AwsCredentialsProvider {
+    private LakeFormationClient lakeFormationClient;
+
+    private String tableArn;
+    private static final List<String> privilegeList = Arrays.asList("SELECT", "INSERT", "DELETE");
```

Comment:
static variables should be above instance variables

Reply:
do we really need to hard code this list? It seems like we can just specify `.permissions(Permission.ALL)` later in the request